felix/labelindex/named_port_index.go

```diff
@@ -775,3 +877,104 @@ func (idx *SelectorAndNamedPortIndex) maybeReportLive() {
 	idx.OnAlive()
 	idx.lastLiveReport = time.Now()
 }
+
+// iterEndpointCandidates iterates over the subset of endpoints that the
+// index says _may_ match the given IP set's selector.  It may produce additional
+// non-matching endpoints (or all endpoints if no optimization is available).
+func (idx *SelectorAndNamedPortIndex) iterEndpointCandidates(ipsetID string, f func(epID any, epData *endpointData)) {
+	sel := idx.ipSetDataByID[ipsetID].selector
+	restrictions := sel.LabelRestrictions()
```

Comment:
I had intended to use the cache in the LabelRestrictionIndex to avoid recalculating but apparently I failed, so let's cache in the selector root instead, where it cannot be bypassed.